File: chess.scala/Move.scala
package chess

abstract sealed class Move {
  val from: Field
  val to: Field
}
case class RegularMove(from: Field, to: Field) extends Move
case class PromotionMove(from: Field, to: Field, figure: Figure) extends Move
case class EnPassantMove(from: Field, to: Field, captured: Field) extends Move
case class CastlingMove(from: Field, to: Field, rookFrom: Field, rookTo: Field) extends Move

File: chess.clj/chess/move.clj
(ns chess.move
  "The `chess.move` namespace contains the functionality for representing
  figure moves.")

(defrecord RegularMove [from to])
(defrecord PromotionMove [from to figure])
(defrecord EnPassantMove [from to captured])
(defrecord CastlingMove [from to from-rook to-rook])

(comment

java clojure.main
(use 'chess.move)

)

